File: src/content/docs/topics/cs/leetcode-150/trees/572-subtree-of-another-tree.py
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def build_tree(vals):
    if not vals:
        return None
    root = TreeNode(vals[0])
    q = [root]
    i = 1
    while q and i < len(vals):
        node = q.pop(0)
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            q.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            q.append(node.right)
        i += 1
    return root

def is_same_tree(p, q):
    pass  # TODO: implement

def is_subtree(root, subRoot):
    pass  # TODO: implement

def _run_tests():
    assert is_subtree(build_tree([3, 4, 5, 1, 2]), build_tree([4, 1, 2])) == True
    assert is_subtree(build_tree([3, 4, 5, 1, 2, None, None, None, None, 0]), build_tree([4, 1, 2])) == False
    t = build_tree([1, 2, 3])
    assert is_subtree(t, t) == True
    assert is_subtree(build_tree([1, 2, 3]), build_tree([2])) == True
    assert is_subtree(build_tree([1, 2, 3]), build_tree([4])) == False
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/trees/572-subtree-of-another-tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
	// TODO: implement
	return false
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	// TODO: implement
	return false
}

func runTests() {
	// [3,4,5,1,2], subRoot=[4,1,2] -> true
	root1 := &TreeNode{Val: 3,
		Left: &TreeNode{Val: 4, Left: &TreeNode{Val: 1}, Right: &TreeNode{Val: 2}},
		Right: &TreeNode{Val: 5},
	}
	sub1 := &TreeNode{Val: 4, Left: &TreeNode{Val: 1}, Right: &TreeNode{Val: 2}}
	assert(isSubtree(root1, sub1) == true)

	// [3,4,5,1,2,null,null,null,null,0], subRoot=[4,1,2] -> false
	root2 := &TreeNode{Val: 3,
		Left: &TreeNode{Val: 4,
			Left:  &TreeNode{Val: 1},
			Right: &TreeNode{Val: 2, Left: &TreeNode{Val: 0}},
		},
		Right: &TreeNode{Val: 5},
	}
	sub2 := &TreeNode{Val: 4, Left: &TreeNode{Val: 1}, Right: &TreeNode{Val: 2}}
	assert(isSubtree(root2, sub2) == false)

	// subRoot is the whole tree
	t := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	sub3 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	assert(isSubtree(t, sub3) == true)

	// single node subroot found
	root3 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	assert(isSubtree(root3, &TreeNode{Val: 2}) == true)

	// single node subroot not found
	assert(isSubtree(root3, &TreeNode{Val: 4}) == false)

	fmt.Println("all tests pass")
}

func main() { runTests() }
